packages/algob/src/builtin-tasks/compile.ts

```diff
@@ -1,13 +1,112 @@
+import type { Algodv2, CompileOut } from "algosdk";
+import { readdirSync, readFileSync, writeFileSync } from 'fs';
+import * as path from 'path';
+import YAML from "yaml";
+
 import { task } from "../internal/core/config/config-env";
-// import { createClient } from "../lib/driver";
-// import { AlgobRuntimeEnv, TaskArguments } from "../types";
+import { parseAlgorandError } from "../internal/core/errors";
+import { assertDir, ASSETS_DIR, CACHE_DIR } from "../internal/core/project-structure";
+import { createClient } from "../lib/driver";
+import type { AlgobRuntimeEnv, ASCCache, Network } from "../types";
 import { TASK_COMPILE } from "./task-names";
+const murmurhash = require('murmurhash'); // eslint-disable-line @typescript-eslint/no-var-requires
+
+const tealExt = ".teal";
 
 export default function (): void {
-  task(TASK_COMPILE, "Compilation task")
-    .setAction(
-      async () => {
-        console.log("compiling......");
-      }
-    );
+  task(TASK_COMPILE, "Compile all TEAL smart contracts")
+    .addFlag("force", "recompile even if the source file didn't change")
+    .setAction(_compile);
+}
+
+export interface TaskArgs {
+  force: boolean
+}
+
+async function _compile ({ force }: TaskArgs, env: AlgobRuntimeEnv): Promise<void> {
+  const op = new CompileOp(env.network);
+  return await compile(force, op);
+}
+
+export async function compile (force: boolean, op: CompileOp): Promise<void> {
+  await assertDir(CACHE_DIR);
+  const cache = readArtifacts(CACHE_DIR);
+
+  for (const f of readdirSync(ASSETS_DIR)) {
+    if (!f.endsWith(tealExt)) { continue; }
+
+    let c = cache.get(f);
+    const [teal, thash] = readTealAndHash(path.join(ASSETS_DIR, f));
+    if (!force && c !== undefined && c.srcHash === thash) {
+      console.log(`smart-contract source "${f}" didn't change, skipping.`);
+      continue;
+    }
+    console.log("compiling", f);
+    c = await compileAndSave(op, new ASCCachePartial(f, teal, thash));
+    cache.set(f, c);
+    const cacheFilename = path.join(CACHE_DIR, f + ".yaml");
+    op.writeFile(cacheFilename, YAML.stringify(c));
+  }
+}
+
+class CompileOp {
+  algocl: Algodv2;
+
+  constructor (n: Network) {
+    this.algocl = createClient(n);
+  }
+
+  async compile (code: string): Promise<CompileOut> {
+    return await this.algocl.compile(code).do();
+  }
+
+  writeFile (filename: string, content: string): void {
```

Comment:
This function doesn't depend on `this.algocl` in any way. I think it's not needed to be in this class.
And then it means that this wrapping class may not be needed. Is it planned to be mocked? I don't see much use for it.

Reply:
Why it should depend on the `algocl` ? The class is a compiler operator and it's responsible for IO. I'm using it to mock it in tests. It's not a `algocl` wrapper. That's why I use the name "compiler operator"

Reply:
If you wouldn't have mocked it in tests then this could be replaced by a static function.